[SR-12912] Fix crash in test targets when accessing Bundle.module

Steven0351 opened the pull request:
Currently, running `swift test` fails when a test target accesses `Bundle.module` during a test (https://forums.swift.org/t/swift-5-3-spm-resources-in-tests-uses-wrong-bundle-path/37051).
This patch uses the bundlePath that already exists instead of crafting the path from `Bundle.main`. I've tested this in both library, executable, and test targets on macOS and it seems to do the trick.

abertelrud commented on Sources/Build/BuildPlan.swift:
```diff
@@ -578,16 +578,13 @@ public final class SwiftTargetBuildDescription {
         extension Foundation.Bundle {
             static var module: Bundle = {
-                let bundlePath = Bundle.main.bundlePath + "/" + "\(bundleBasename)"
+                let bundlePath = "\(bundlePath)"
```
Wouldn't this hardcode the absolute path into each artifact (also non-tests)?

Steven0351 replied:
I'm not familiar with how `bundlePath` is derived so that is possible. I wasn't able to confirm this on an iOS device itself as Xcode won't let me install it on my test device with a custom toolchain. Is it possible to check if it's compiling a test target?

neonichu replied:
Be aware that the Xcode integration of packages does not use SwiftPM from the toolchain, you will only see an effect of your changes in commandline SwiftPM.

abertelrud replied:
Right, but in this case, wouldn't this also change the path that's built into a package target that's linked into an executable, meaning that if the executable and its associated resource bundles are later moved, the bundles can't be found?

Steven0351 replied:
> Right, but in this case, even when just building with `swift build`, wouldn't this also change the path that's built into a package target that's linked into an executable, meaning that if the executable and its associated resource bundles are later moved, the bundles can't be found?

That seems likely, when I tested I did move the executable but not the resource bundle. Would it make sense to conditionally check if the current Bundle.main path is located in an Xcode path and fallback to the hard coded path? 

@neonichu I'm guessing that Xcode has its own method for computing Bundle.module? I had gone down a rabbit hole of trying to run my tests with xcodebuild on my package, but had other issues altogether there.

Steven0351 commented on Sources/Build/BuildPlan.swift:
```diff
@@ -577,14 +577,16 @@ public final class SwiftTargetBuildDescription {
         guard let bundlePath = self.bundlePath else { return }
-
+      
+        let pathToModule = isTestTarget ? bundlePath : "Bundle.main.bundlePath/\(bundlePath.basename)"
```
I had just realized this is not actually going to word, I've made some changes locally and building a new toolchain before pushing, but tentatively this has been re-written to 
```swift
let pathToModule = isTestTarget ? bundlePath : "Bundle.main.bundlePath + \"/\" + \(bundlePath.basename)"
```

Steven0351 commented:
I've converted this to a Draft PR because this is going to take a bit of consideration. To address the concern raised by @abertelrud, I updated the generation of `Bundle.module` to this:
```swift
let pathToModule = isTestTarget ? #""\#(bundlePath.pathString)""# : #"Bundle.main.bundlePath + "/" + "\#(bundlePath.basename)""#
      
        let stream = BufferedOutputByteStream()
        stream <<< """
        import class Foundation.Bundle

        extension Foundation.Bundle {
            static var module: Bundle = {
                let bundlePath = \(pathToModule)
                guard let bundle = Bundle(path: bundlePath) else {
                    fatalError("could not load resource bundle: \\(bundlePath)")
                }
                return bundle
            }()
        }
        """
```
However, when a target that is _being_ tested calls `Bundle.module` to access a resource during a test, the test fails with the same error:
Fatal error: could not load resource bundle: <path to xcode developer tools>/usr/bin/Foo_Foo.bundle: file <path to derived sources>/resource_bundle_accessor.swift, line 7

Steven0351 commented:
I think I've come up with a reasonable solution: we first check if the module bundle is accessible by appending to the bundle name to Bundle.main, then we fallback to the build directory, and if that fails we bail out with a fatal error.

Jeehut commented:
@Steven0351 Did you already find time to implement your "reasonable solution" so we can get this merged before final release of Xcode 12? Feel free to ping me if I can help here as the bug is really blocking a proper CI configuration for any open source repo using resources right now ...

Steven0351 commented:
@Jeehut I had committed what I thought was a "reasonable" solution, but there has been no feedback on those commits after I marked them ready for review. Bundle.module is generated at build time, so my workaround has the original bundle, but falls back to the build path in the event that the preferred bundle is not available (which should only be the case when a target is under test). I've been using my fork to run tests for macOS Swift packages without issue.

MaxDesiatov commented:
I can trigger CI and/or request reviews from relevant people, is this ready for CI or review at all?

Steven0351 commented:
@MaxDesiatov It's ready for review. I've run the tests locally when I originally patched and validated that my macOS Swift PM projects with bundle resources were able to both be tested and run (along with being moved around the file system without breaking non-test behavior).

MaxDesiatov commented:
Would you be able to add corresponding automated tests to the SwiftPM test suite, so that it could be tested on CI?

Steven0351 commented:
@MaxDesiatov I could give it a go but I'll have to dig into the test suite to see how this part of the codebase is tested. I could probably write something that validates the values that are generated, but I didn't notice anything in the current test suite that would enable validating the properties at runtime. Do you have a suggested approach there?

MaxDesiatov commented:
Did you check the integration test suite with test cases [such as this one](https://github.com/apple/swift-package-manager/blob/master/IntegrationTests/Tests/IntegrationTests/SwiftPMTests.swift)?

Steven0351 commented:
@MaxDesiatov I had never noticed the IntegrationsTest sub project since it's not part of the main project. I ran into a couple of issues there: it uses the existing swift toolchain for that test (at least when building with Swift PM). I added the integration test into Tests/Commands/TestToolTest.swift (probably not the best place for that), but to have a passing test I need to be able access the swift-test binary compiled for the build. However, I did add the failing integration test in the IntegrationTests project so if anyone is so inclined to validate themselves locally, but this would probably fail swift-ci. Any suggestions on where this particular test should actually live would be welcome.

@neonichu Would you be able to review/throw in your opinion here as well?

abertelrud commented:
@swift-ci please smoke test

abertelrud commented on Sources/Build/BuildPlan.swift:
```diff
@@ -578,19 +578,22 @@ public final class SwiftTargetBuildDescription {
+                let buildPath = "\(bundlePath.pathString)"
+
+                let preferredBundle = Bundle(path: mainPath)
+                let fallBackBundle = Bundle(path: buildPath)
```
Since this could be a little expensive, it might be worth instantiating this second Bundle only if the first one can't be found, especially since at deployment time for the non-test products, the first one should always be found.

Steven0351 replied:
That's fair; I've inlined the instantiation of this in the ternery expression per this feedback

abertelrud commented:
Thanks for making the change!

abertelrud commented:
@swift-ci please test

neonichu commented:
Looks like the only failure is from an accidental full test run, so this can be merged.

ffried commented:
Any chance for this to make it into Swift 5.3?

ffried commented:
@MaxDesiatov Thank you!

odrobnik commented:
Forgive my ignorance, how can one tell in which Swift version this fix will be included? Swift 5.3, as of Xcode 12 still has the original implementation as evidenced with the shorter error message.

MaxDesiatov commented:
I saw some folks mentioned it's in Xcode 12.2 Beta 2, but I haven't verified it myself

Steven0351 commented:
I can verify it is available in Xcode 12.2 Beta 2.

djbe commented:
Unfortunately it's not in Xcode 12.1 though 😢 So we'll have to wait until 12.2 is released.
